logging: swo: Add option to set custom reference frequency

nordic-krch opened the pull request:
Some devices may use different reference than cpu clock. Add
Kconfig option to provide custom frequency reference details.

Fixes #31058.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: swo: Add option to set custom reference frequency

Some devices may use different reference than cpu clock. Add
support for using swo-ref-frequency property when present.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

mnkp commented:
Thanks, this issue is long overdue. I believe however that we should fix it using DT property and not Kconfig option. If the DT property (e.g. `swo-frequency`) is attached to the CPU node then SWO driver should use it, otherwise it should use CPU frequency.

mnkp commented:
@nordic-krch @carlescufi I don't want to block this PR as I haven't been involved in Zephyr development recently, however SWO frequency is a hardware property that needs to be configured during system boot. It really should be defined via DTS, not Kconfig. In a few months during next round of Kconfig cleanup we will have another PR that will move this setting to DTS. Let's do it right from the beginning.

nordic-krch commented:
@mnkp you are right, i've been busy with others stuff recently and could not get back to that, Soon i should have some time.

nordic-krch committed:
dts: bindings: cpu: Add cortex-m common properties

Added file with common properties for cortex-m cores.
Added optional swo-ref-frequency property.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
dts: arm: nordic: Add swo-ref-frequency property

Added swo-req-frequency property to nordic SoC's

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
soc: arm: nordic: Indicate SWO presence in nrf52, nrf53 and nrf91

Add HAS_SWO to nordic SoC series.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
I've added `dts/bindings/cpu/arm,cortex-m.yaml` which is included by all cortex-mx cpu's. It contains optional `swo-ref-freq` property. Added this property to nordic devices (also added `HAS_SWO`). Modified logging swo backend to use this property if present. @mnkp is that what you had in mind?

mnkp commented on dts/bindings/cpu/arm,cortex-m.yaml:
```diff
@@ -0,0 +1,12 @@
+include: cpu.yaml
+
+properties:
+    swo-ref-freq:
```
What about changing the name of the property to `swo-ref-frequency`? Since we have `clock-frequency` for CPU clock.

mnkp reviewed (commented):
Thanks, that looks good. I would remove the first commit "logging: swo: Add option to set custom reference frequency" though. You nuke it anyway, less clutter.

nordic-krch commented:
@mnkp renamed property and squashed changes in `log_backend_swo` to single commit.

mnkp reviewed (commented):
The "logging: swo: Add option to set custom reference frequency" commit message needs to be updated. Still mentions Kconfig option.

nordic-krch commented on subsys/logging/log_backend_swo.c:
```diff
@@ -36,18 +36,25 @@
-#define SWO_FREQ_DIV  (SWO_FREQ / CONFIG_LOG_BACKEND_SWO_FREQ_HZ)
+
+#define SWO_FREQ_DIV \
+	(SWO_REF_FREQ / CONFIG_LOG_BACKEND_SWO_FREQ_HZ)
```
maybe there should be an error when target frequency cannot be matched?

mnkp replied:
That would be an option, however a legal output frequency can be a rational number. Let's say - in the example above - user wants to set up SWO trace with divider equal to 3, to achieve output frequency of 10 MHz / 3 = 3.3333 MHz.

nordic-krch replied:
added compilation error when target frequency cannot be achieved. Expecting that there will be up to 1 hz difference between reference and div * target to accommodate for fractional number.

mnkp replied:
This will not work if we divide by a different number, e.g. 7. For other combinations of dividend and divisor the error can be larger. It's safer to round `SWO_FREQ_DIV` to the nearest integer using the formula I proposed.

If you are worried that the formula is not readable, maybe we should create a macro that would round to the nearest integer and put it in utils?

nordic-krch replied:
Used your formula, can you revisit?

mnkp reviewed (commented):
Looks good. Let's only change the commit message of "logging: swo: Add option to set custom reference frequency " commit. It sill mentions Kconfig.

nordic-krch commented:
@mnkp commit message fixed.
